### PR TITLE
Feature/deterministic nfr runs

### PR DESCRIPTION
**Pull request by improbable-danny**
Deterministic movement of NPCs and simplayers in NFR runs

### CODE REVIEW

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 1]
```diff
@@ -9,6 +9,9 @@
 #include "GameFramework/Character.h"
 #include "Engine/World.h"
 #include "Kismet/GameplayStatics.h"
+#include "AIController.h"
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:12]: Sort these includes

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 2]
```diff
@@ -45,8 +48,33 @@ ABenchmarkGymGameMode::ABenchmarkGymGameMode()
 
 	NPCSToSpawn = 0;
 	SecondsTillPlayerCheck = 15.0f * 60.0f;
+}
 
-	RNG.Initialize(123456); // Ensure we can do deterministic runs
+void ABenchmarkGymGameMode::GenerateTestScenarioLocations()
+{
+	const float ROAM_RADIUS = 7500.0f;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:55]: No need for caps, and could you add a comment how we get to this value (half actor ncd)

File: Game/Source/GDKTestGyms/GDKTestGyms.Build.cs [hunk 3]
```diff
@@ -22,7 +22,11 @@ public GDKTestGyms(ReadOnlyTargetRules Target) : base(Target)
 				"OnlineSubsystemUtils",
 				"PhysXVehicles",
 				"SpatialGDK",
-				"ReplicationGraph"
+				"ReplicationGraph",
+                "AIModule",
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/GDKTestGyms.Build.cs:26]: Spacing is off here

File: Game/Source/GDKTestGyms/GDKTestGyms.Build.cs [hunk 4]
```diff
@@ -22,7 +22,11 @@ public GDKTestGyms(ReadOnlyTargetRules Target) : base(Target)
...
+                "NavigationSystem"
 			});
-	}
+
+        OptimizeCode = CodeOptimization.Never;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/GDKTestGyms.Build.cs:30]: Remove this

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 5]
```diff
@@ -45,8 +48,33 @@ ABenchmarkGymGameMode::ABenchmarkGymGameMode()
...
+	{
+		FRandomStream PlayerStream;
+		PlayerStream.Initialize(123456); // Ensure we can do deterministic runs
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:58]: What about making this seed a factor of the ExpectedPlayers? So they are still deterministic, but player 1 doesn't always run through the same points.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 6]
```diff
@@ -45,8 +48,33 @@ ABenchmarkGymGameMode::ABenchmarkGymGameMode()
...
+		for (int i = 0; i < ExpectedPlayers; i++)
+		{
+			FVector PointA = PlayerStream.VRand()*ROAM_RADIUS;
+			FVector PointB = PlayerStream.VRand()*ROAM_RADIUS;
+			PointA.Z = PointB.Z = 0.0f;
+			PlayerRunPoints.Push(RunPoints{ PointA, PointB });
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:64]: nit - use emplace and just pass the objects directly - ie PlayerRunPoints.Emplace({ PointA, PointB });

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 7]
```diff
@@ -87,10 +117,11 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
 	{
 		if (NPCSToSpawn > 0)
 		{
-			int32 Cluster = (--NPCSToSpawn) % NumPlayerClusters;
+			int32 NPCIndex = --NPCSToSpawn;
+			int32 Cluster = (NPCIndex) % NumPlayerClusters;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:121]:
```suggestion
			int32 Cluster = NPCIndex % NumPlayerClusters;
```

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 8]
```diff
@@ -87,10 +117,11 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
...
 			int32 SpawnPointIndex = Cluster * PlayerDensity;
 			const AActor* SpawnPoint = SpawnPoints[SpawnPointIndex];
-			SpawnNPC(SpawnPoint->GetActorLocation());
+			SpawnNPC(SpawnPoint->GetActorLocation(), NPCRunPoints[NPCIndex % NPCRunPoints.Num()]);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:124]: Curious - why do we do this in tick, and not BeginPlay ?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 9]
```diff
@@ -309,8 +360,11 @@ AActor* ABenchmarkGymGameMode::FindPlayerStart_Implementation(AController* Playe
 	PlayerIdToSpawnPointMap.Add(PlayerUniqueID, ChosenSpawnPoint);
 
 	UE_LOG(LogBenchmarkGym, Log, TEXT("Spawning player %d at %s."), PlayerUniqueID, *ChosenSpawnPoint->GetActorLocation().ToString());
+	
+	AIControlledPlayers.Push(Player);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:364]: Didn't we have some concern at some point that FindPlayerStart could be called multiple times for a single player?

Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:364]: Also this says AIControlledPlayers, but wouldn't this get hit for normal clients? Should we check to see if this is a sim player before attempting to add it?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 10]
```diff
@@ -110,6 +141,26 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
 				}
 			}
 		}
+
+		for (int i = AIControlledPlayers.Num() - 1; i >= 0; i--)
+		{
+			if (AIControlledPlayers[i] == nullptr)
+			{
+				AIControlledPlayers.RemoveAt(i);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:149]: RemoveAtSwap is quicker. Might also be nicer to just have a single array with both the Player and Idx.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 11]
```diff
@@ -110,6 +141,26 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
...
+				AIControllerPlayerRunIndexes.RemoveAt(i);
+				continue;
+			}
+
+			ACharacter* Player = AIControlledPlayers[i]->GetCharacter();
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:154]: Perhaps rename to Character? Player seems a little overused.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 12]
```diff
@@ -110,6 +141,26 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
...
+			int Info = AIControllerPlayerRunIndexes[i];
+			if (UDeterministicBlackboardValues* Blackboard = (UDeterministicBlackboardValues*)Player->FindComponentByClass(UDeterministicBlackboardValues::StaticClass()))
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:156]: Unreal cast

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 13]
```diff
@@ -110,6 +141,26 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
...
+			{
+				auto& Points = PlayerRunPoints[Info % PlayerRunPoints.Num()];
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:158]: No autos

Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:158]: Why the %? Shouldn't there be as many runpoints as there are players?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 14]
```diff
@@ -273,19 +324,19 @@ void ABenchmarkGymGameMode::SpawnNPC(const FVector& SpawnLocation)
...
+	FVector FixedSpawnLocation = SpawnLocation*0.9f + RunLocations.A*0.1f; // Move slightly away from the spawn location
 	UE_LOG(LogBenchmarkGym, Log, TEXT("Spawning NPC at %s"), *SpawnLocation.ToString());
 	FActorSpawnParameters SpawnInfo{};
 	//SpawnInfo.Owner = this;
 	//SpawnInfo.Instigator = NULL;
 	SpawnInfo.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
-	World->SpawnActor<APawn>(NPCPawnClass->GetDefaultObject()->GetClass(), RandomSpawn, FRotator::ZeroRotator, SpawnInfo);
+	APawn* Pawn = World->SpawnActor<APawn>(NPCPawnClass->GetDefaultObject()->GetClass(), FixedSpawnLocation, FRotator::ZeroRotator, SpawnInfo);
+	checkf(Pawn, TEXT("Pawn failed to spawn at %s"), *FixedSpawnLocation.ToString());
+	//SetBlackboardValuesForAIController(Pawn->GetController(), RunLocations);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:335]: Remove commented out code

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 15]
```diff
@@ -273,19 +324,19 @@ void ABenchmarkGymGameMode::SpawnNPC(const FVector& SpawnLocation)
...
+	
+	UDeterministicBlackboardValues* Comp = (UDeterministicBlackboardValues*)Pawn->FindComponentByClass(UDeterministicBlackboardValues::StaticClass());
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:337]: Unreal cast

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+#include "BehaviorTree/BlackboardComponent.h"
+#include "NavigationSystem.h"
+
+DEFINE_LOG_CATEGORY(LogDeterministicBlackboardValues);
+
+// Sets default values for this component's properties
+UDeterministicBlackboardValues::UDeterministicBlackboardValues()
+{
+	bWantsInitializeComponent = true;
+	//PrimaryComponentTick.bWantsInitializeComponent = true;
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:16]: Commented code

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+void UDeterministicBlackboardValues::InitializeComponent()
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:19]: Seems redundant

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+	//PrimaryComponentTick.bWantsInitializeComponent = true;
+}
+
+void UDeterministicBlackboardValues::InitializeComponent() 
+{
+	Super::InitializeComponent();
+}
+
+// Called every frame
+void UDeterministicBlackboardValues::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:25]: Likewise

File: Game/Source/GDKTestGyms/Public/DeterministicBlackboardValues.h [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+{
+	GENERATED_BODY()
+
+public:	
+	// Sets default values for this component's properties
+	UDeterministicBlackboardValues();
+
+protected:
+	FTimerHandle TimerHandle;
+	FVector RunPointA; // Points to run betweeen
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/DeterministicBlackboardValues.h:22]: Might be worth GameMode and this actor sharing the same Struct.

File: Game/Source/GDKTestGyms/Public/DeterministicBlackboardValues.h [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+	void InitializeComponent() override;
+
+	// Called every frame
+	void TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction) override;
+
+	void ApplyValues();
+
+	UFUNCTION(Client, Reliable)
+	void SetValues(const FVector& A, const FVector& B);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Public/DeterministicBlackboardValues.h:34]: Function could be more descriptive

Comment (m-samiec) [reply Game/Source/GDKTestGyms/Public/DeterministicBlackboardValues.h:34]: Even prefix it with `Client` to indicate it's an RPC

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+		UE_LOG(LogDeterministicBlackboardValues, Log, TEXT("Could not find an AI controller."));
+//#endif
+	}
+}
+void UDeterministicBlackboardValues::SetValues_Implementation(const FVector& A, const FVector& B)
+{
+	RunPointA = A;
+	RunPointB = B;
+
+	GetWorld()->GetTimerManager().SetTimer(TimerHandle, this, &UDeterministicBlackboardValues::ApplyValues, 1.0f, true);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:80]: Why the delay? Please add a comment explaining.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 8 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:124]: I had a theory that there were some flakes caused by a spike on the server but I don't think this was the cause. We could move it back but I suppose spawning 300+ actors with nav queries is still quite heavy

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+		Blackboard->SetValueAsVector(TargetA, LocA.Location);
+		Blackboard->SetValueAsVector(TargetB, LocB.Location);
+
+		UE_LOG(LogDeterministicBlackboardValues, Log, TEXT("Setting points to run between as %s and %s for AI controller %s"), *RunPointA.ToString(), *RunPointB.ToString(), *Controller->GetName());
+		GetWorld()->GetTimerManager().ClearTimer(TimerHandle);
+	}
+
+	else
+	{
+//#if !WITH_EDITOR
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:70]: Commented out code

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+// Called every frame
+void UDeterministicBlackboardValues::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);
+}
+
+void UDeterministicBlackboardValues::ApplyValues()
+{
+	APawn* P = Cast<APawn>(GetOwner());
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:32]:
```suggestion
	APawn* Pawn = Cast<APawn>(GetOwner());
```

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+
+		FVector WorldLocation = ((AActor*)Controller)->GetActorLocation(); // GetWorldLocation?
+
+	// 	FVector RunLocationA = WorldLocation + Locations.A;
+	// 	FVector RunLocationB = WorldLocation + Locations.B;
+
+		UNavigationSystemV1* NavSys = UNavigationSystemV1::GetCurrent(GetWorld());
+
+		FNavLocation LocA;
+		checkf(NavSys->GetRandomPointInNavigableRadius(WorldLocation + RunPointA, 100.0f, LocA), TEXT("Was not able to find point to nav to."));
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:56]: Can't put this in a check or it'll get optimised away in none-checking builds.
Also 100.f seems a little magic.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 9 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:364]: The map test above should deal with multiple calls to FindPlayerStart for the same controller

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 13 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:158]: There should be, although there isn't anything to prevent spawning too many players either on a local build or in the cloud. Perhaps there should be some warning log here as well

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp (hunk 24 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp]: It's totally magic, I added some tolerance to allow for some distance between the point and the nav mesh surface - 1m seemed to do the trick without breaking the determinism

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 8 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:125]: Ah okay, yeah could be cleaned up, but not necessary.

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 25]
```diff
@@ -110,6 +142,30 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
 				}
 			}
 		}
+
+		for (int i = AIControlledPlayers.Num() - 1; i >= 0; i--)
+		{
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:147]: Can't this logic just be simplified to if controller and character are good, call `ClientSetBlackboardAILocations`, then once whole array is processed, empty it?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 26]
```diff
@@ -273,19 +329,16 @@ void ABenchmarkGymGameMode::SpawnNPC(const FVector& SpawnLocation)
...
 	}
 
-	const float RandomSpawnOffset = 600.0f;
-	FVector RandomOffset = RNG.GetUnitVector()*RandomSpawnOffset;
-	if (RandomOffset.Z < 0.0f)
-	{
-		RandomOffset.Z = -RandomOffset.Z;
-	}
-	FVector RandomSpawn = SpawnLocation + RandomOffset;
+	FVector FixedSpawnLocation = SpawnLocation*0.9f + BlackboardValues.TargetAValue*0.1f; // Move slightly away from the spawn location
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:332]: Why the change?

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp [hunk 27]
```diff
@@ -297,10 +350,15 @@ AActor* ABenchmarkGymGameMode::FindPlayerStart_Implementation(AController* Playe
 		return Super::FindPlayerStart_Implementation(Player, IncomingName);
 	}
 
+	if (Player == nullptr) // Work around for load balancing passing nullptr Player
+	{
+		return SpawnPoints[PlayersSpawned % SpawnPoints.Num()];
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:355]: Should PlayersSpawned be incremented here? Or is this the instance where it's called twice?

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+		Blackboard->SetValueAsVector(BlackboardValues.TargetAName, LocA);
+		Blackboard->SetValueAsVector(BlackboardValues.TargetBName, LocB);
+
+		UE_LOG(LogDeterministicBlackboardValues, Log, TEXT("Setting points to run between as %s and %s for AI controller %s"), *LocA.Location.ToString(), *LocB.Location.ToString(), *Controller->GetName());
+		GetWorld()->GetTimerManager().ClearTimer(TimerHandle);
+	}
+	else
+	{
+#if !WITH_EDITOR
+		UE_LOG(LogDeterministicBlackboardValues, Log, TEXT("Could not find an AI controller."));
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:57]: Does this log spam currently?

Comment (m-samiec) [reply Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:57]: Include retry in msg

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+		FVector WorldLocation = ((AActor*)Controller)->GetActorLocation();
+
+		UNavigationSystemV1* NavSys = UNavigationSystemV1::GetCurrent(GetWorld());
+
+		constexpr float Tolerance = 100.0f; // A tolerance to allow the point to snap to the nav mesh surfaces.
+
+		FNavLocation LocA;
+		{
+			FVector LocalRandomPoint = WorldLocation + BlackboardValues.TargetAValue;
+			bool Result = NavSys->GetRandomPointInNavigableRadius(LocalRandomPoint, Tolerance, LocA);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:37]:
```suggestion
			bool bResult = NavSys->GetRandomPointInNavigableRadius(LocalRandomPoint, Tolerance, LocA);
```

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 27 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:355]: Yeah, the load balancing code calls this with Player = nullptr, guess it's just a temporary workaround for now! So it'll return the same point here and below in the second call

File: Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp (hunk 26 above)
Comment (improbable-danny) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:332]: Just so this logic is also deterministic, I doubt the random offset would cause many problems I suppose it could have 2 players run into each other and get stuck

Comment (m-samiec) [reply Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp:322]: Oh, because the movement goes -
communal spawn location > point a > point b > point a (etc.)
Perhaps we should spawn them directly at point a?

File: Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp [hunk 30]
```diff
@@ -0,0 +1,93 @@
...
+
+void UDeterministicBlackboardValues::ApplyValues() // Repeats until the Component is added 
+{
+	APawn* Pawn = Cast<APawn>(GetOwner());
+	AController* Controller = Pawn->GetController();
+
+	if (Controller == nullptr)
+	{
+#if !WITH_EDITOR
+		UE_LOG(LogDeterministicBlackboardValues, Log, TEXT("Pawn contorller is not set, retrying in 1 second."));
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/Private/DeterministicBlackboardValues.cpp:20]:
```suggestion
		UE_LOG(LogDeterministicBlackboardValues, Log, TEXT("Pawn controller is not set, retrying in 1 second."));
```
